src/Heap.js

```diff
@@ -51,34 +77,61 @@ export default class Heap {
     }
 
     buildMinHeap() { 
-        const lastParent = Math.floor((this.heap.length - 2) / 2);
+        const lastParent = Math.floor((this.heap.length - 1) / 2);
 
         for (let i = lastParent; i >= 0; i--) {
             this.minHeapify(i);
         }
     }
 
     checkHeap(position) {
-        while (position> 0) {
-            position = Math.floor((position - 2) / 2);
+        while (position > 0) {
+            position = Math.floor((position - 1) / 2);
             this.minHeapify(position);
         }
     }
 
     insert(node, key) {
         this.heap.push({node, key});
         this._nodeIndices.set(node, this.heap.length - 1);
-        this.checkHeap(this.heap.length - 1);
+        this.filterUp(this.heap.length - 1);
     }
 
     compare(a, b) {
         if (a && b) {
-            return a.key - b.key;
+            if (Array.isArray(a) && Array.isArray(b)) {
+                if (a[0] > b[0] || (a[0] === b[0] && a[1] > b[1])) {
+                    return 1;
+                }
+
+                return -1;
+            }
+
+            return a - b;
         }
 
         return 0;
     }
 
+    remove(node) {
+        const index = this._nodeIndices.get(node);
+        if (index == null) {
+            return;
+        }
+        if (index === this.heap.length - 1) {
+            this._nodeIndices.delete(this.heap.pop().node);
+            return;
+        }
+
+        this.swap(index, this.heap.length - 1);
+        this._nodeIndices.delete(this.heap.pop().node);
```

Comment:
since this is the same line as earlier, i wonder if we could be clever about this and DRY?

Reply:
ex.
```js
    remove(node) {
        const index = this._nodeIndices.get(node);
        if (index == null) {
            return;
        }

        const lastIndex = this.heap.length - 1;

        // Move the item to the bottom and then remove it.
        this.swap(index, lastIndex);
        const last = this.heap.pop();
        this._nodeIndices.delete(last.node);

        // If it was already at the bottom, we're done.
        if (index === lastIndex) {
            return;
        }

        const parentIndex = Math.floor((index - 1) / 2);
        if (index === 0 || this.compare(this.heap[parentIndex].key, this.heap[index].key) < 0) {
            // If we're at the top, or the parent is smaller, we should sift down.
            this.minHeapify(index);
        } else {
            // Otherwise, sift up.
            this.filterUp(index);
        }
    }
```

working through this, i noticed two areas where we could be clever-er:
- `swap` can do an early return if `a === b`
- `compare` can take in indices instead of keys, saving us a *lot* of boilerplate every time we need to call it

Reply:
(i may have gotten those comments wrong but you know it better than i do so 😊)